### PR TITLE
[BEAM-10670] Make Read use SDF by default. Override in runners.

### PR DESCRIPTION
**Pull request by lukecwik**
We add a PTransform override which converts to a new primitive read transform updating the existing read translation logic to only apply to this new primitive read and not to Read.Bounded/Read.Unbounded.

A good chunk of this change is fixing up test expectations since the expansion is always using SDF now.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by lukecwik** [comment]
R: @TheNeuralBit

**Comment by lukecwik** [comment]
CC: @kennknowles This pushes the logic down from Read into each runner to override with a primitive read.

@TheNeuralBit This is the other half of the change which exposed the issue with Samza and SDF.

**Comment by lukecwik** [comment]
Run Flink ValidatesRunner Batch

**Comment by lukecwik** [comment]
Run Flink ValidatesRunner

**Comment by lukecwik** [comment]
Run Twister2 ValidatesRunner

**Comment by lukecwik** [comment]
Run Spark ValidatesRunner

**Comment by lukecwik** [comment]
Run Dataflow ValidatesRunner

**Comment by lukecwik** [comment]
Run Samza ValidatesRunner

**Comment by lukecwik** [comment]
Run Samza ValidatesRunner

**Comment by lukecwik** [comment]
Run Java_Examples_Dataflow_Java11 PreCommit

**Comment by lukecwik** [comment]
Run Flink ValidatesRunner

**Review by kennknowles — COMMENTED** [review]
Nice! The Samza failures all appear to be related. (different reuse of same assertion)

File: runners/samza/src/main/java/org/apache/beam/runners/samza/SamzaRunner.java
Comment (kennknowles) [inline runners/samza/src/main/java/org/apache/beam/runners/samza/SamzaRunner.java:110]: To clarify: this change makes it so instead of rejecting primitive reads the function is a noop?

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ReadTranslator.java
Comment (kennknowles) [inline runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ReadTranslator.java:57]: I was trying to figure out where the failure came from and I would assume here but I don't understand why.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/SamzaRunner.java
Comment (lukecwik) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/SamzaRunner.java:110]: Its a noop if the experiments aren't set. If the experiments are set then `Read.Bounded` and `Read.Unbounded` get converted into `PrimitiveBoundedRead` and `PrimitiveUnboundedRead` respectively.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ReadTranslator.java
Comment (lukecwik) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ReadTranslator.java:57]: This is some pre-existing bug in the Samza runner that I'm exposing as all the other validates runner tests pass that used to pass and now they are using SDF based Reads everywhere (which covers Create as well).

Comment (lukecwik) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/translation/ReadTranslator.java:57]: When debugging locally I can see that the PAssert executes once and gets the correct input and then it executes another time only containing `"finished"` (coming from the finishBundle on the TestFn) which implies that the PAssert trigger is occurring multiple times causing the verifier to fail on a subsequent bundle.

**Comment by lukecwik** [comment]
Run Flink ValidatesRunner

**Comment by lukecwik** [comment]
Run Twister2 ValidatesRunner

**Comment by lukecwik** [comment]
Run Spark ValidatesRunner

**Comment by lukecwik** [comment]
Run Dataflow ValidatesRunner

**Comment by lukecwik** [comment]
Run Samza ValidatesRunner

**Comment by lukecwik** [comment]
Run Java_Examples_Dataflow_Java11 PreCommit

**Comment by lukecwik** [comment]
Run Flink ValidatesRunner

**Comment by lukecwik** [comment]
Run Flink ValidatesRunner

**Comment by lukecwik** [comment]
R: @kennknowles If you got the time it would be greatly appreciated since you did a first pass.

**Comment by lukecwik** [comment]
Run Java PreCommit

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/SplittableParDo.java
Comment (kennknowles) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/SplittableParDo.java:653]: I believe this should be controlled by the runner choosing to invoke the method, not by a global flag. It can have the same status as other runner-internal overrides, like GBK via GBKO.

If you really believe it should be a flag, it should be the runner that reads the flag and decides what to do. This utility library should not change its behavior based on pipeline options. Only runners should opt in to particular behaviors.

File: runners/samza/src/main/java/org/apache/beam/runners/samza/SamzaRunner.java
Comment (kennknowles) [reply runners/samza/src/main/java/org/apache/beam/runners/samza/SamzaRunner.java:110]: For a runner that previously rejected all primitive reads, isn't it better to leave that runner alone and still reject all primitive reads?

(here and elsewhere)

**Review by kennknowles — COMMENTED** [review]
I did a first pass. I think big picture there are two things:

1. Without unfolding all the code not in the code review, I am not already familiar with the expected level of SDF support of all the runners. Some of them seem to previously reject non-SDF reads but now could end up having non-SDF reads.
2. I do not think runners core construction should be manipulated by global flags, but each runner should decide what to do with the flag, if anything. If a runner has a well-used non-portable variant that is not going away any time soon, then it can still use the portable pipeline and just reject any with non-Java UDFs. If a runner only supports SDF in some modes but only supports Read in others, or wants to translate Read specially for efficiency (or any other reason it wants to - it can do whatever it wants) then it can do that.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/SplittableParDo.java
Comment (lukecwik) [reply runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/SplittableParDo.java:653]: I can do that but it seems like it will lead to possible inconsistency across runners around how this choice is made. Currently CHANGES.md says to get the old behavior use this flag and it would be unfortunate to say "consult the runner documentation as to how to get the old behavior". Also the intent is for the old behavior to be removed in a few releases so that we don't have to support it anymore.

**Comment by lukecwik** [comment]
> I did a first pass. I think big picture there are two things:
> 
> 1. Without unfolding all the code not in the code review, I am not already familiar with the expected level of SDF support of all the runners. Some of them seem to previously reject non-SDF reads but now could end up having non-SDF reads.

BEAM-10670 is about migrating all runners to use SDF first and only SDF. All runners that supported BoundedSource/UnboundedSource support SDF to the level necessary to execute the *BoundedSource as SDF wrappers. The only ones that don't are Spark (underway in #12603) and Dataflow runner v1 which would have a degraded experience.

> 2. I do not think runners core construction should be manipulated by global flags, but each runner should decide what to do with the flag, if anything. If a runner has a well-used non-portable variant that is not going away any time soon, then it can still use the portable pipeline and just reject any with non-Java UDFs. If a runner only supports SDF in some modes but only supports Read in others, or wants to translate Read specially for efficiency (or any other reason it wants to - it can do whatever it wants) then it can do that.

I can do that but it seems like it will lead to possible inconsistency across runners around how this choice is made. Currently CHANGES.md says to get the old behavior use this flag and it would be unfortunate to say "consult the runner documentation as to how to get the old behavior". 

After a few releases, the intent is that this logic is removed from runners-core and everyone except for Dataflow v1 runner always use the SDF wrapper and we move the necessary logic within runners-core to the Dataflow module deleting the remainder (the PrimitiveBoundedRead override).

**Comment by lukecwik** [comment]
Run Flink ValidatesRunner

**Comment by lukecwik** [comment]
Run Java_Examples_Dataflow_Java11 PreCommit

**Comment by lukecwik** [comment]
Run Samza ValidatesRunner

**Comment by lukecwik** [comment]
Run Dataflow ValidatesRunner

**Comment by lukecwik** [comment]
Run Twister2 ValidatesRunner

**Comment by lukecwik** [comment]
Run Spark ValidatesRunner

**Comment by kennknowles** [comment]
> @kennknowles
> 
> > I did a first pass. I think big picture there are two things:
> > 
> > 1. Without unfolding all the code not in the code review, I am not already familiar with the expected level of SDF support of all the runners. Some of them seem to previously reject non-SDF reads but now could end up having non-SDF reads.
> 
> [BEAM-10670](https://issues.apache.org/jira/browse/BEAM-10670) is about migrating all runners to use SDF first and only SDF. All runners that supported BoundedSource/UnboundedSource support SDF to the level necessary to execute the BoundedSource/UnboundedSource as SDF wrappers. The only ones that don't are Spark (underway in #12603) and Dataflow runner v1 which would have a degraded experience in batch and would break pipeline update in streaming. This was a focus of mine over the past ~6 weeks.

Do they have equivalent performance? My architectural idea is that we can have expansion that looks like:

```
Read(BoundedSource) {
  Impulse -> SDF
}
```

preserving all the bounded/unbounded source data but having a portable expansion. "Portable" meaning that all non-primitives (like Read) are expanded to primitives (SDF). But there is no reason a runner has to support SDF at all, if they can recognize the composite and implement it directly.

So my basic idea would be to have all such runners continue to use their existing execution paths, unless the runner authors/maintainers believe whatever SDF support they have is good enough. FWIW I don't think this is my invention. I thought this was the plan to seamlessly support Read and SDF in the same portable pipeline, with runners picking and choosing what they can execute CC @xinyuiscool @iemejia @mxm

> > 1. I do not think runners core construction should be manipulated by global flags, but each runner should decide what to do with the flag, if anything. If a runner has a well-used non-portable variant that is not going away any time soon, then it can still use the portable pipeline and just reject any with non-Java UDFs. If a runner only supports SDF in some modes but only supports Read in others, or wants to translate Read specially for efficiency (or any other reason it wants to - it can do whatever it wants) then it can do that.
> 
> I can do that but it seems like it will lead to possible inconsistency across runners around how this choice is made. Currently CHANGES.md says to get the old behavior use this flag and it would be unfortunate to say "consult the runner documentation as to how to get the old behavior".

There should not be behavior differences. The expansion above should have identical PCollection contents whether the runner implements Read or SDF. And the runner should make sure that pipeline maintain their performance by using the best translation they can. Users should not need to be aware of this, except perhaps via debugging interfaces that expose the expansion.
 
> After a few releases, the intent is that this logic is removed from runners-core and everyone except for Dataflow v1 runner always use the SDF wrapper and we move the necessary logic within runners-core to the Dataflow module deleting the remainder (the PrimitiveBoundedRead override).

My position on runners-core and runners-core-construction is that these are just libraries with collections of useful functionality for implementing a runner. Neither one should ever be user-visible. Deleting dead code from either library is nice-to-have cleanup. If any runner wants to keep the code it can be copied into the runner.

**Comment by kennknowles** [comment]
*note that "recognize the composite and implement it directly" means they both recognize the URN _and_ recognize that they can execute the Java-based `BoundedSource`/`UnboundedSource` directly. There may be some complexity around the portable process boundary providing diamond-dep protection, but that may be worth it for more direct translation to the underlying system, and that is a decision for runner authors/maintainers/users.

**Comment by kennknowles** [comment]
> > > @kennknowles
> > > > I did a first pass. I think big picture there are two things:
> > > > 
> > > > 1. Without unfolding all the code not in the code review, I am not already familiar with the expected level of SDF support of all the runners. Some of them seem to previously reject non-SDF reads but now could end up having non-SDF reads.
> > > 
> > > 
> > > [BEAM-10670](https://issues.apache.org/jira/browse/BEAM-10670) is about migrating all runners to use SDF first and only SDF. All runners that supported BoundedSource/UnboundedSource support SDF to the level necessary to execute the BoundedSource/UnboundedSource as SDF wrappers. The only ones that don't are Spark (underway in #12603) and Dataflow runner v1 which would have a degraded experience in batch and would break pipeline update in streaming. This was a focus of mine over the past ~6 weeks.
> > 
> > 
> > Do they have equivalent performance? My architectural idea is that we can have expansion that looks like:
> 
> Nexmark did show improvements ([#12519 (comment)](https://github.com/apache/beam/pull/12519#issuecomment-680278642))
> 
> > ```
> > Read(BoundedSource) {
> >   Impulse -> SDF
> > }
> > ```
> > 
> > 
> > preserving all the bounded/unbounded source data but having a portable expansion. "Portable" meaning that all non-primitives (like Read) are expanded to primitives (SDF). But there is no reason a runner has to support SDF at all, if they can recognize the composite and implement it directly.
> > So my basic idea would be to have all such runners continue to use their existing execution paths, unless the runner authors/maintainers believe whatever SDF support they have is good enough. FWIW I don't think this is my invention. I thought this was the plan to seamlessly support Read and SDF in the same portable pipeline, with runners picking and choosing what they can execute CC @xinyuiscool @iemejia @mxm
> 
> I brought up the migration to using SDF to power BoundedSource/UnboundedSource on [dev@](https://lists.apache.org/thread.html/r1ba6fe6ac2bd2b28aa7ef31f0d87ad716fc878f2515085fdbc275333%40%3Cdev.beam.apache.org%3E). If you would like to reconsider the migration path it would make sense to continue this discussion on the original thread.
> 
> > > > 1. I do not think runners core construction should be manipulated by global flags, but each runner should decide what to do with the flag, if anything. If a runner has a well-used non-portable variant that is not going away any time soon, then it can still use the portable pipeline and just reject any with non-Java UDFs. If a runner only supports SDF in some modes but only supports Read in others, or wants to translate Read specially for efficiency (or any other reason it wants to - it can do whatever it wants) then it can do that.
> > > 
> > > 
> > > I can do that but it seems like it will lead to possible inconsistency across runners around how this choice is made. Currently CHANGES.md says to get the old behavior use this flag and it would be unfortunate to say "consult the runner documentation as to how to get the old behavior".
> > 
> > 
> > There should not be behavior differences. The expansion above should have identical PCollection contents whether the runner implements Read or SDF. And the runner should make sure that pipeline maintain their performance by using the best translation they can. Users should not need to be aware of this, except perhaps via debugging interfaces that expose the expansion.
> 
> The purpose of having the fallback is to give users an opt-out mechanism during the migration should they hit issues. I don't believe our runners have comprehensive testing of pipelines at scale enough to state that the migration will be perfect.
> 
> > > After a few releases, the intent is that this logic is removed from runners-core and everyone except for Dataflow v1 runner always use the SDF wrapper and we move the necessary logic within runners-core to the Dataflow module deleting the remainder (the PrimitiveBoundedRead override).
> > 
> > 
> > My position on runners-core and runners-core-construction is that these are just libraries with collections of useful functionality for implementing a runner. Neither one should ever be user-visible. Deleting dead code from either library is nice-to-have cleanup. If any runner wants to keep the code it can be copied into the runner.
> 
> I disagree. As someone who went through all our runners changing a core piece, dead code would only make Apache Beam harder to maintain.

"core" is just a name because I/we didn't have a clever name for it. Both libraries were created for this purpose and really shouldn't be seen as more than a utility library. They are not the definition of any part of the Beam model, just helpers.

**Comment by lukecwik** [comment]
> > > > @kennknowles
> > > > > I did a first pass. I think big picture there are two things:
> > > > > 
> > > > > 1. Without unfolding all the code not in the code review, I am not already familiar with the expected level of SDF support of all the runners. Some of them seem to previously reject non-SDF reads but now could end up having non-SDF reads.
> > > > 
> > > > 
> > > > [BEAM-10670](https://issues.apache.org/jira/browse/BEAM-10670) is about migrating all runners to use SDF first and only SDF. All runners that supported BoundedSource/UnboundedSource support SDF to the level necessary to execute the BoundedSource/UnboundedSource as SDF wrappers. The only ones that don't are Spark (underway in #12603) and Dataflow runner v1 which would have a degraded experience in batch and would break pipeline update in streaming. This was a focus of mine over the past ~6 weeks.
> > > 
> > > 
> > > Do they have equivalent performance? My architectural idea is that we can have expansion that looks like:
> > 
> > 
> > Nexmark did show improvements ([#12519 (comment)](https://github.com/apache/beam/pull/12519#issuecomment-680278642))
> > > ```
> > > Read(BoundedSource) {
> > >   Impulse -> SDF
> > > }
> > > ```
> > > 
> > > 
> > > preserving all the bounded/unbounded source data but having a portable expansion. "Portable" meaning that all non-primitives (like Read) are expanded to primitives (SDF). But there is no reason a runner has to support SDF at all, if they can recognize the composite and implement it directly.
> > > So my basic idea would be to have all such runners continue to use their existing execution paths, unless the runner authors/maintainers believe whatever SDF support they have is good enough. FWIW I don't think this is my invention. I thought this was the plan to seamlessly support Read and SDF in the same portable pipeline, with runners picking and choosing what they can execute CC @xinyuiscool @iemejia @mxm
> > 
> > 
> > I brought up the migration to using SDF to power BoundedSource/UnboundedSource on [dev@](https://lists.apache.org/thread.html/r1ba6fe6ac2bd2b28aa7ef31f0d87ad716fc878f2515085fdbc275333%40%3Cdev.beam.apache.org%3E). If you would like to reconsider the migration path it would make sense to continue this discussion on the original thread.
> > > > > 1. I do not think runners core construction should be manipulated by global flags, but each runner should decide what to do with the flag, if anything. If a runner has a well-used non-portable variant that is not going away any time soon, then it can still use the portable pipeline and just reject any with non-Java UDFs. If a runner only supports SDF in some modes but only supports Read in others, or wants to translate Read specially for efficiency (or any other reason it wants to - it can do whatever it wants) then it can do that.
> > > > 
> > > > 
> > > > I can do that but it seems like it will lead to possible inconsistency across runners around how this choice is made. Currently CHANGES.md says to get the old behavior use this flag and it would be unfortunate to say "consult the runner documentation as to how to get the old behavior".
> > > 
> > > 
> > > There should not be behavior differences. The expansion above should have identical PCollection contents whether the runner implements Read or SDF. And the runner should make sure that pipeline maintain their performance by using the best translation they can. Users should not need to be aware of this, except perhaps via debugging interfaces that expose the expansion.
> > 
> > 
> > The purpose of having the fallback is to give users an opt-out mechanism during the migration should they hit issues. I don't believe our runners have comprehensive testing of pipelines at scale enough to state that the migration will be perfect.
> > > > After a few releases, the intent is that this logic is removed from runners-core and everyone except for Dataflow v1 runner always use the SDF wrapper and we move the necessary logic within runners-core to the Dataflow module deleting the remainder (the PrimitiveBoundedRead override).
> > > 
> > > 
> > > My position on runners-core and runners-core-construction is that these are just libraries with collections of useful functionality for implementing a runner. Neither one should ever be user-visible. Deleting dead code from either library is nice-to-have cleanup. If any runner wants to keep the code it can be copied into the runner.
> > 
> > 
> > I disagree. As someone who went through all our runners changing a core piece, dead code would only make Apache Beam harder to maintain.
> 
> "core" is just a name because I/we didn't have a clever name for it. Both libraries were created for this purpose and really shouldn't be seen as more than a utility library. They are not the definition of any part of the Beam model, just helpers.

I didn't mean "core" as in the modules with the name "core" in them but the "core" concept of how data is ingested into a pipeline by a runner.

**Comment by kennknowles** [comment]
> #12519 (comment)



> > > > > @kennknowles
> > > > > > I did a first pass. I think big picture there are two things:
> > > > > > 
> > > > > > 1. Without unfolding all the code not in the code review, I am not already familiar with the expected level of SDF support of all the runners. Some of them seem to previously reject non-SDF reads but now could end up having non-SDF reads.
> > > > > 
> > > > > 
> > > > > [BEAM-10670](https://issues.apache.org/jira/browse/BEAM-10670) is about migrating all runners to use SDF first and only SDF. All runners that supported BoundedSource/UnboundedSource support SDF to the level necessary to execute the BoundedSource/UnboundedSource as SDF wrappers. The only ones that don't are Spark (underway in #12603) and Dataflow runner v1 which would have a degraded experience in batch and would break pipeline update in streaming. This was a focus of mine over the past ~6 weeks.
> > > > 
> > > > 
> > > > Do they have equivalent performance? My architectural idea is that we can have expansion that looks like:
> > > 
> > > 
> > > Nexmark did show improvements ([#12519 (comment)](https://github.com/apache/beam/pull/12519#issuecomment-680278642))
> > > > ```
> > > > Read(BoundedSource) {
> > > >   Impulse -> SDF
> > > > }
> > > > ```
> > > > 
> > > > 
> > > > preserving all the bounded/unbounded source data but having a portable expansion. "Portable" meaning that all non-primitives (like Read) are expanded to primitives (SDF). But there is no reason a runner has to support SDF at all, if they can recognize the composite and implement it directly.
> > > > So my basic idea would be to have all such runners continue to use their existing execution paths, unless the runner authors/maintainers believe whatever SDF support they have is good enough. FWIW I don't think this is my invention. I thought this was the plan to seamlessly support Read and SDF in the same portable pipeline, with runners picking and choosing what they can execute CC @xinyuiscool @iemejia @mxm
> > > 
> > > 
> > > I brought up the migration to using SDF to power BoundedSource/UnboundedSource on [dev@](https://lists.apache.org/thread.html/r1ba6fe6ac2bd2b28aa7ef31f0d87ad716fc878f2515085fdbc275333%40%3Cdev.beam.apache.org%3E). If you would like to reconsider the migration path it would make sense to continue this discussion on the original thread.
> > > > > > 1. I do not think runners core construction should be manipulated by global flags, but each runner should decide what to do with the flag, if anything. If a runner has a well-used non-portable variant that is not going away any time soon, then it can still use the portable pipeline and just reject any with non-Java UDFs. If a runner only supports SDF in some modes but only supports Read in others, or wants to translate Read specially for efficiency (or any other reason it wants to - it can do whatever it wants) then it can do that.
> > > > > 
> > > > > 
> > > > > I can do that but it seems like it will lead to possible inconsistency across runners around how this choice is made. Currently CHANGES.md says to get the old behavior use this flag and it would be unfortunate to say "consult the runner documentation as to how to get the old behavior".
> > > > 
> > > > 
> > > > There should not be behavior differences. The expansion above should have identical PCollection contents whether the runner implements Read or SDF. And the runner should make sure that pipeline maintain their performance by using the best translation they can. Users should not need to be aware of this, except perhaps via debugging interfaces that expose the expansion.
> > > 
> > > 
> > > The purpose of having the fallback is to give users an opt-out mechanism during the migration should they hit issues. I don't believe our runners have comprehensive testing of pipelines at scale enough to state that the migration will be perfect.
> > > > > After a few releases, the intent is that this logic is removed from runners-core and everyone except for Dataflow v1 runner always use the SDF wrapper and we move the necessary logic within runners-core to the Dataflow module deleting the remainder (the PrimitiveBoundedRead override).
> > > > 
> > > > 
> > > > My position on runners-core and runners-core-construction is that these are just libraries with collections of useful functionality for implementing a runner. Neither one should ever be user-visible. Deleting dead code from either library is nice-to-have cleanup. If any runner wants to keep the code it can be copied into the runner.
> > > 
> > > 
> > > I disagree. As someone who went through all our runners changing a core piece, dead code would only make Apache Beam harder to maintain.
> > 
> > 
> > "core" is just a name because I/we didn't have a clever name for it. Both libraries were created for this purpose and really shouldn't be seen as more than a utility library. They are not the definition of any part of the Beam model, just helpers.
> 
> I didn't mean "core" as in the modules with the name "core" in them but the "core" concept of how data is ingested into a pipeline by a runner.

OK. I am only referring to the module name. The modules are just piles of functionality that a runner can use as they please. The thing I don't like is having a global user-controlled flag that tweaks runner behavior when the runner may not even be aware. Basically moving a flag from the SDK to the runner is better but still not great.

**Review by kennknowles — APPROVED** [review]
The fixes here seem valuable and it is an improvement so LGTM. I do want to follow up and remove the use of globals/flags in the core libs which are intended to be utility libraries. But happy to have this merged to make good progress.

**Pull request merged**
